src/ameba/rule/lint/typos.cr

```diff
@@ -0,0 +1,97 @@
+module Ameba::Rule::Lint
+  # A rule that reports typos found in source files.
+  #
+  # NOTE: Needs [typos](https://github.com/crate-ci/typos) CLI tool.
+  # NOTE: See the chapter on [false positives](https://github.com/crate-ci/typos#false-positives).
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Lint/Typos:
+  #   Enabled: true
+  #   BinPath: ~
+  #   FailOnError: false
+  # ```
+  class Typos < Base
+    properties do
+      description "Reports typos found in source files"
+      bin_path nil.as(String?)
+      fail_on_error false
+    end
+
+    record Typo,
+      path : String,
+      typo : String,
+      corrections : Array(String),
+      location : {Int32, Int32},
+      end_location : {Int32, Int32} do
+      def self.parse(str)
+        issue = JSON.parse(str)
+
+        return unless issue["type"] == "typo"
+
+        typo = issue["typo"].as_s
+        corrections = issue["corrections"].as_a.map(&.as_s)
+
+        return if typo.empty? || corrections.empty?
+
+        path = issue["path"].as_s
+        line_no = issue["line_num"].as_i
+        col_no = issue["byte_offset"].as_i + 1
+        end_col_no = col_no + typo.size - 1
+
+        new(path, typo, corrections,
+          {line_no, col_no}, {line_no, end_col_no})
+      end
+    end
+
+    MSG = "Typo found: %s -> %s"
+
+    protected class_getter bin_path : String? {
+      Process.find_executable("typos")
+    }
+
+    @[YAML::Field(ignore: true)]
+    protected getter bin_path : String {
+      @bin_path ||
+        self.class.bin_path ||
+        raise RuntimeError.new "Could not find `typos` executable"
+    }
+
+    protected def typos_from(source : Source)
+      status = Process.run(bin_path, args: ["--format", "json", source.path],
+        output: output = IO::Memory.new,
+      )
+      return if status.success?
+
+      ([] of Typo).tap do |typos|
+        # NOTE: `--format json` is actually JSON Lines (`jsonl`)
+        output.to_s.each_line do |line|
+          Typo.parse(line).try { |typo| typos << typo }
+        end
+      end
+    end
+
+    def test(source : Source)
+      typos = typos_from(source)
+      typos.try &.each do |typo|
+        message = MSG % {
+          typo.typo, typo.corrections.join(" / "),
+        }
+
+        # FIXME
+        # if typo.corrections.size == 1
+        #   issue_for typo.location, typo.end_location, message do |corrector|
+        #     corrector.replace(typo.location, typo.end_location, typo.corrections.first)
+        #   end
+        # else
+        #   issue_for typo.location, typo.end_location, message
+        # end
```

Comment:
what's the problem with corrections? I think it would be very nice having it for this specific rule

Reply:
Sure thing, that's why I left it for further work. ATM it doesn't work for some reason - try uncommenting the code and check for yourself. Looks like a corrector bug, or some other strange creature - perhaps you'll be able to spot the error 👀

Reply:
I ended up having `Error: Infinite loop in my_file.cr caused by Lint/Typos`, however, was not able to find the cause. Still looking

Reply:
yup, same here

Reply:
@FnControlOption Any idea what might be going on?

Reply:
Hm… Could it be that there’s an issue being reported that doesn’t get fixed by autocorrect?

Reply:
None other then the typos, it seems.

Reply:
Right. Do you mind logging the original source at `typo.location` and the expected correction `typo.corrections.first` and pasting the output here?

Reply:
Sure, just a sec.

Reply:
```cr
def tpos_frm(userl)
  pp :foo
end
```

```cr
Ameba::Rule::Lint::Typos::Typo(
  @typo="tpos",
  @corrections=["typos"],
  @location={1, 5},
  @end_location={1, 8},
  @path="bug.cr"
)
```

Reply:
It seems like off-by-one type of error. I'm not sure where though.

Reply:
It magically started to work... 🤯 @veelenga Could you confirm?

Reply:
Hm, yes, it is fixed for me as well.